---
title: "Single-cell profiling identifies pre-existing CD19-negative subclones in a B-ALL patient with CD19-negative relapse after CAR-T therapy : CaSpER analysis"
author: "Delphine Potier"
date: "`r format(Sys.time(), '%d %B, %Y')`"
always_allow_html: yes
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: hide
  pdf_document: 
    toc: true
    toc_depth: 3
    highlight: zenburn
editor_options: 
  chunk_output_type: console
---

Authors : Tracy Rabilloud\*, Delphine Potier\*, Saran Pankew, Mathis Nozais, Marie Loosveld§, Dominique Payet-Bornet§

\* Equal contribution

§ Corresponding authors


PMID: to come

Raw data and intermediate data matrices are available in SRA/GEO (SRP269742 / GSE153697) 

Docker images and intermediate seurat object are available in zenodo.
Any questions on this analysis, please contact Delphine Potier

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r library variable and object loading, include=FALSE}
# Load libraries
library(Seurat)
library(ggplot2)
library(plotly)
library(ggpubr)
library(GenomicRanges)
library(CaSpER)

# Load variables & objects
## Path to the CaSpER analysis folder
WD_PATH <- "<WORKING_DIR>/3_CaSpER_analysis/"
## Path to cluster 0, cluster 1 (T1-CD19neg cells excluded) and clusters 2,3,4,5 BAFs
BAF_PATH <- "<WORKING_DIR>/3_CaSpER_analysis/output/cluster_subset/" 
## Path to cluster 0 & cluster 1 T1-CD19neg cells BAFs
OTHER_BAF_PATH <- "<WORKING_DIR>/3_CaSpER_analysis/output/cluster_subset/Other_BAF/"

load(file = paste0(WD_PATH,"/data/hg38_cytoband.rda"))
load("<WORKING_DIR>/1_Seurat_analysis/output/Seurat_final_BALL-CART.Robj")
```

```{r Get the gene expression matrix, include=FALSE}
log.ge <- as.matrix(Seurat@assays$RNA@counts)
genes <- rownames(log.ge)
```

```{r Get annotation, include=FALSE}
#annotation <- generateAnnotation(id_type="hgnc_symbol", genes=genes, centromere=centromere, ishg19 = F)
#save(annotation, file = paste0(WD_PATH, "annotation_CaSpER.Robj"))
load(file = paste0(WD_PATH,"/data/annotation_CaSpER.Robj"))
```

```{r formatting data for CaSpER, include=FALSE}
log.ge <- log.ge[match( annotation$Gene,rownames(log.ge)), ]
rownames(log.ge) <- annotation$Gene
log.ge <- log2(log.ge +1)
#Create a new metadata field for bafgroups
## Cluster 0 & 1 contains tumoral cell at T1 and T2, while the orther clusters contains physiological cells (we grouped them together)
Seurat@meta.data$bafgroups <- "nothing"
Seurat@meta.data[row.names(subset(Seurat@meta.data, RNA_snn_res.0.1 == "0")),]$bafgroups = "cluster0_baf"
Seurat@meta.data[row.names(subset(Seurat@meta.data, RNA_snn_res.0.1 == "1")),]$bafgroups = "cluster1_baf"
Seurat@meta.data[row.names(subset(Seurat@meta.data, RNA_snn_res.0.1 == "2")),]$bafgroups = "clusters_2_3_4_5_baf"
Seurat@meta.data[row.names(subset(Seurat@meta.data, RNA_snn_res.0.1 == "3")),]$bafgroups = "clusters_2_3_4_5_baf"
Seurat@meta.data[row.names(subset(Seurat@meta.data, RNA_snn_res.0.1 == "4")),]$bafgroups = "clusters_2_3_4_5_baf"
Seurat@meta.data[row.names(subset(Seurat@meta.data, RNA_snn_res.0.1 == "5")),]$bafgroups = "clusters_2_3_4_5_baf"

## Set Seurat objet idents to the bafgroups
Idents(Seurat) <- "bafgroups"

## Put physiological cells in "control"
control <- names(Idents(Seurat))[Idents(Seurat) %in% c("clusters_2_3_4_5_baf")]

# Load bafextract results (here cluster0_baf.snp, cluster1_baf.snp and clusters_2_3_4_5_baf.snp)

loh <- readBAFExtractOutput(path = BAF_PATH, sequencing.type = "single-cell")

names(loh) <- gsub(".snp","", names(loh))
loh.name.mapping <- data.frame (loh.name= Seurat@meta.data$bafgroups , sample.name=colnames(log.ge))
names(loh)
```


```{r creating CaSpER object, include=FALSE}
# Create CaSpER object
object <- CreateCasperObject(raw.data=log.ge,loh.name.mapping=loh.name.mapping, sequencing.type="single-cell",
cnv.scale=3, loh.scale=3,
expr.cutoff=0.1, filter="median", matrix.type="normalized",
annotation=annotation, method="iterative", loh=loh,
control.sample.ids=control, cytoband=cytoband_hg38)
```


```{r running CaSpER, include=FALSE}
## runCaSpER
final.objects <- runCaSpER(object, removeCentromere=T, cytoband=cytoband, method="iterative")
```

```{r extractLargeScaleEvents, include=FALSE}
## summarize large scale events
finalChrMat <- extractLargeScaleEvents(final.objects, thr=1) # In the CaSpER paper figures, a threshold of 1 correspond to 9 (indeed, in the function mergeScalesDel is divided by the object length which 9 so 9/9 = 1 to be compared to this threshold:  "finalChrMat[(mergeScalesDel/length(final.objects)) >= thr] <- (-1)" )

```

# 9q deletion

## {.tabset}
### Stringeant gamma threshold

* gamma threshold = 9

In the CaSpER paper figure 4C (see https://www.nature.com/articles/s41467-019-13779-x), a Ɣ threshold of 9 correspond to a thr of 1 in the extractLargeScaleEvents() function.
Indeed, in the function, mergeScalesDel variable is divided by the object length which is 9 (9/9 = 1) 
("finalChrMat[(mergeScalesDel/length(final.objects)) >= thr] <- (-1)")



#### UMAP
```{r plot 9q deletion, fig.width = 5, fig.height = 7}
# Plot 9q deletion score
## Clones of interest : Get barcodes for T1-CD19neg clones located in the T1 tumoral area (Cluster 0) for highlighting
Idents(Seurat) <- "RNA_snn_res.0.1"
HTO_T1CD19n_in_tumor <- intersect(row.names(subset(Seurat@meta.data, hash.ID == "T1-CD19neg")), WhichCells(Seurat, slot = "RNA_snn_res.0.1", idents = c("0","1")))

## Add 9q status in the metadata
Seurat@meta.data$chr9q <- "nothing"
Seurat@meta.data$chr9q <- finalChrMat[,"9q"]

## Draw UMAP
### Get the coordinates & values
Umap<-data.frame(
UMAP_1 = Seurat@reductions$umap@cell.embeddings[,1],
UMAP_2= Seurat@reductions$umap@cell.embeddings[,2],
cnv= Seurat@meta.data$chr9q)

HTO= Seurat@meta.data$hash.ID
Max=max(Seurat@meta.data$chr9q)
Min=min(Seurat@meta.data$chr9q)

ggplot(Umap,aes(x=UMAP_1,y=UMAP_2))+geom_point(aes(color=as.factor(cnv),shape=HTO),size = 0.8)+ scale_color_manual(name = "Chr9q", labels = c("Deletion","Neutral","Amplification"),values=c("blue", "grey", "orange"))+geom_point(data = Umap[HTO_T1CD19n_in_tumor,],aes(x=UMAP_1,y=UMAP_2, fill = as.factor(cnv)),shape = 21,size = 3, color = "red")+ scale_fill_manual(values=c("blue", "grey", "orange")) 
```


Interactive UMAP
```{r 9q deletion interactive umap, fig.width = 5, fig.height = 7} 
### Plot an interactive UMAP
ggplotly(ggplot(Umap,aes(x=UMAP_1,y=UMAP_2))+geom_point(aes(color=as.factor(cnv),shape=HTO),size = 0.6)+ scale_color_manual(name = "Chr9q", labels = c("Deletion","Neutral","Amplification"),values=c("blue", "grey", "orange"))+geom_point(data = Umap[HTO_T1CD19n_in_tumor,],aes(x=UMAP_1,y=UMAP_2, fill = as.factor(cnv)),shape = 21,size = 2, color = "red")+ scale_fill_manual(values=c("blue", "grey", "orange")))

```




#### 9q deletion summary 

1) Number of 9q large scale event by group of cells

```{r 9q deletion table} 
C0_chr9q_del_counts <- table(factor(Seurat@meta.data[which(Seurat@meta.data$RNA_snn_res.0.1 == "0"),]$chr9q, levels = -1:1))

C1_chr9q_del_counts <- table(Seurat@meta.data[which(Seurat@meta.data$RNA_snn_res.0.1 == "1"),]$chr9q)

C2345_chr9q_del_counts <- table(Seurat@meta.data[which(Seurat@meta.data$RNA_snn_res.0.1 == "2" | Seurat@meta.data$RNA_snn_res.0.1 == "3" | Seurat@meta.data$RNA_snn_res.0.1 == "4" | Seurat@meta.data$RNA_snn_res.0.1 =="5"),]$chr9q)

 
chr9q_table <- matrix(data = c(C0_chr9q_del_counts,C1_chr9q_del_counts,C2345_chr9q_del_counts), nrow = 3,ncol = 3 )
rownames(chr9q_table) <- c("Deletion","Neutral","Amplification")
colnames(chr9q_table) <- c("Cluster 0","Cluster 1","Clusters 2, 3, 4, 5")

chr9q_table
```



2) Number of 9q large scale event for T1CD19neg cells in tumoral clusters (0 & 1)
```{r HTO_T1CD19n_in_tumor 9q deletion table} 
Seurat20c <- subset(Seurat, cells = HTO_T1CD19n_in_tumor)

HTO_T1CD19n_in_tumor_chr9q_del_counts <- table(factor(Seurat20c@meta.data[,]$chr9q, levels = -1:1))

HTO_T1CD19n_in_tumor_chr9q_table <- matrix(data = c(HTO_T1CD19n_in_tumor_chr9q_del_counts), nrow = 3,ncol = 1 )
rownames(HTO_T1CD19n_in_tumor_chr9q_table) <- c("Deletion","Neutral","Amplification")
colnames(HTO_T1CD19n_in_tumor_chr9q_table) <- c("T1CD19neg cells in tumoral clusters (0 & 1)")

HTO_T1CD19n_in_tumor_chr9q_table
```



3) Proportions of 9q large scale event by group of cells
```{r 9q deletion table percentage}
round(prop.table(chr9q_table,2)*100,2)
```



### Relaxed gamma threshold

* gamma threshold = 5


```{r plot 9q deletion 2, fig.width = 5, fig.height = 7}
## summarize large scale events 
finalChrMat2 <- extractLargeScaleEvents (final.objects, thr=0.65) # less stringeant, increase the sensitivity but deacrease specificity, by default thr is est at 0.5

# Plot 9q deletion score
Seurat@meta.data$chr9q <- "nothing"
Seurat@meta.data$chr9q <- finalChrMat2[,"9q"]

  Umap<-data.frame(
    UMAP_1 = Seurat@reductions$umap@cell.embeddings[,1],
    UMAP_2= Seurat@reductions$umap@cell.embeddings[,2],
    cnv= Seurat@meta.data$chr9q
  )
  HTO= Seurat@meta.data$hash.ID
  Max=max(Seurat@meta.data$chr9q)
  Min=min(Seurat@meta.data$chr9q)

#Non interactive UMAP
ggplot(Umap,aes(x=UMAP_1,y=UMAP_2))+geom_point(aes(color=as.factor(cnv),shape=HTO),size = 0.8)+ scale_color_manual(name = "Chr9q", labels = c("Deletion","Neutral","Amplification"),values=c("blue", "grey", "orange"))+geom_point(data = Umap[HTO_T1CD19n_in_tumor,],aes(x=UMAP_1,y=UMAP_2, fill = as.factor(cnv)),shape = 21,size = 3, color = "red")+ scale_fill_manual(values=c("blue", "grey", "orange"))      
```


Interactive UMAP
```{r 9q deletion interactive umap 2, fig.width = 5, fig.height = 7} 
#Interactive UMAP
ggplotly(ggplot(Umap,aes(x=UMAP_1,y=UMAP_2))+geom_point(aes(color=as.factor(cnv),shape=HTO),size = 0.6)+ scale_color_manual(name = "Chr9q", labels = c("Deletion","Neutral","Amplification"),values=c("blue", "grey", "orange"))+geom_point(data = Umap[HTO_T1CD19n_in_tumor,],aes(x=UMAP_1,y=UMAP_2, fill = as.factor(cnv)),shape = 21,size = 2, color = "red")+ scale_fill_manual(values=c("blue", "grey", "orange")))
```

#### 9q deletion summary 

1) Number of 9q large scale event by group of cells
```{r 9q deletion table relax} 
C0_chr9q_del_counts <- table(factor(Seurat@meta.data[which(Seurat@meta.data$RNA_snn_res.0.1 == "0"),]$chr9q, levels = -1:1))

C1_chr9q_del_counts <- table(Seurat@meta.data[which(Seurat@meta.data$RNA_snn_res.0.1 == "1"),]$chr9q)

C2345_chr9q_del_counts <- table(Seurat@meta.data[which(Seurat@meta.data$RNA_snn_res.0.1 == "2" | Seurat@meta.data$RNA_snn_res.0.1 == "3" | Seurat@meta.data$RNA_snn_res.0.1 == "4" | Seurat@meta.data$RNA_snn_res.0.1 =="5"),]$chr9q)

 
chr9q_table <- matrix(data = c(C0_chr9q_del_counts,C1_chr9q_del_counts,C2345_chr9q_del_counts), nrow = 3,ncol = 3 )
rownames(chr9q_table) <- c("Deletion","Neutral","Amplification")
colnames(chr9q_table) <- c("Cluster 0","Cluster 1","Clusters 2, 3, 4, 5")

chr9q_table
```



2) Number of 9q large scale event for T1CD19neg cells in tumoral clusters (0 & 1)
```{r HTO_T1CD19n_in_tumor 9q deletion table relax} 
Seurat20c <- subset(Seurat, cells = HTO_T1CD19n_in_tumor)

HTO_T1CD19n_in_tumor_chr9q_del_counts <- table(factor(Seurat20c@meta.data[,]$chr9q, levels = -1:1))

HTO_T1CD19n_in_tumor_chr9q_table <- matrix(data = c(HTO_T1CD19n_in_tumor_chr9q_del_counts), nrow = 3,ncol = 1 )
rownames(HTO_T1CD19n_in_tumor_chr9q_table) <- c("Deletion","Neutral","Amplification")
colnames(HTO_T1CD19n_in_tumor_chr9q_table) <- c("T1CD19neg cells in tumoral clusters (0 & 1)")

HTO_T1CD19n_in_tumor_chr9q_table
```



3) Proportions of 9q large scale event by group of cells
```{r 9q deletion table percentage relax}
round(prop.table(chr9q_table,2)*100,2)
```



# Check specific BAF
```{r}
# Load bafextract results (here "cluster0_T1CD19neg_baf_mincov10_2_05", "cluster0_T1CD19neg_baf_mincov10", "cluster0_T1CD19neg_baf_mincov6" & "cluster0_T1CD19neg_baf")
loh_T1CD19n_in_tumor <- readBAFExtractOutput(path = OTHER_BAF_PATH, sequencing.type = "single-cell")
names(loh_T1CD19n_in_tumor) <- gsub(".snp","", names(loh_T1CD19n_in_tumor))
```



## rs10900809
```{r BAF}
obj <- final.objects[[9]]

#Check for some specific BAF
## rs10900809
snp132490630_c0 <- obj@loh$cluster0_baf[which(obj@loh$cluster0_baf$chr == "5" & obj@loh$cluster0_baf$position == "132490630"),]

snp132490630_c1 <- obj@loh$cluster1_baf[which(obj@loh$cluster1_baf$chr == "5" & obj@loh$cluster1_baf$position == "132490630"),]

snp132490630_c2_3_4_5 <- obj@loh$clusters_2_3_4_5_baf[which(obj@loh$clusters_2_3_4_5_baf$chr == "5" & obj@loh$clusters_2_3_4_5_baf$position == "132490630"),]

snp_freq_table <- matrix(data = c(snp132490630_c0$alt,snp132490630_c1$alt,snp132490630_c2_3_4_5$alt,snp132490630_c0$ref,snp132490630_c1$ref,snp132490630_c2_3_4_5$ref,snp132490630_c0$coverage,snp132490630_c1$coverage,snp132490630_c2_3_4_5$coverage,snp132490630_c0$baf,snp132490630_c1$baf,snp132490630_c2_3_4_5$baf), nrow = 3,ncol = 4 )

rownames(snp_freq_table) <- c("Cluster 0","Cluster 1","Clusters 2,3,4,5")
colnames(snp_freq_table) <- c("Alt","Ref","Coverage","BAF")
snp_freq_table

snp132490630_tumT1CD19neg <- loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6[which(loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$chr == "5" & loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$position == "132490630"),]

rownames(snp132490630_tumT1CD19neg) <- "Tumoral T1CD19negative cells"
snp132490630_tumT1CD19neg[,3:6]
```



## rs4705403
```{r BAFrs4705403}
snp150000930_c0 <- obj@loh$cluster0_baf[which(obj@loh$cluster0_baf$chr == "5" & obj@loh$cluster0_baf$position == "150000930"),]

snp150000930_c1 <- obj@loh$cluster1_baf[which(obj@loh$cluster1_baf$chr == "5" & obj@loh$cluster1_baf$position == "150000930"),]

snp150000930_c2_3_4_5 <- obj@loh$clusters_2_3_4_5_baf[which(obj@loh$clusters_2_3_4_5_baf$chr == "5" & obj@loh$clusters_2_3_4_5_baf$position == "150000930"),]



snp_freq_table <- matrix(data = c(snp150000930_c0$alt,snp150000930_c1$alt,snp150000930_c2_3_4_5$alt,snp150000930_c0$ref,snp150000930_c1$ref,snp150000930_c2_3_4_5$ref,snp150000930_c0$coverage,snp150000930_c1$coverage,snp150000930_c2_3_4_5$coverage,snp150000930_c0$baf,snp150000930_c1$baf,snp150000930_c2_3_4_5$baf), nrow = 3,ncol = 4 )

rownames(snp_freq_table) <- c("Cluster 0","Cluster 1","Clusters 2,3,4,5")
colnames(snp_freq_table) <- c("Alt","Ref","Coverage","BAF")
snp_freq_table

### loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6[which(loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$chr == "5" & loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$position == "150402064"),]

print("Less than 6 reads at chr5:150000930 for tumoral T1CD19neg cells")
```



## rs1056400
```{r BAFrs1056400}
snp150402064_c0 <- obj@loh$cluster0_baf[which(obj@loh$cluster0_baf$chr == "5" & obj@loh$cluster0_baf$position == "150402064"),]

snp150402064_c1 <- obj@loh$cluster1_baf[which(obj@loh$cluster1_baf$chr == "5" & obj@loh$cluster1_baf$position == "150402064"),]

snp150402064_c2_3_4_5 <- obj@loh$clusters_2_3_4_5_baf[which(obj@loh$clusters_2_3_4_5_baf$chr == "5" & obj@loh$clusters_2_3_4_5_baf$position == "150402064"),]

snp_freq_table <- matrix(data = c(snp150402064_c0$alt,snp150402064_c1$alt,snp150402064_c2_3_4_5$alt,snp150402064_c0$ref,snp150402064_c1$ref,snp150402064_c2_3_4_5$ref,snp150402064_c0$coverage,snp150402064_c1$coverage,snp150402064_c2_3_4_5$coverage,snp150402064_c0$baf,snp150402064_c1$baf,snp150402064_c2_3_4_5$baf), nrow = 3,ncol = 4 )

rownames(snp_freq_table) <- c("Cluster 0","Cluster 1","Clusters 2,3,4,5")
colnames(snp_freq_table) <- c("Alt","Ref","Coverage","BAF")
snp_freq_table


snp150402064_tumT1CD19neg <- loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6[which(loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$chr == "5" & loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$position == "150402064"),]
rownames(snp150402064_tumT1CD19neg) <- "Tumoral T1CD19negative cells"
snp150402064_tumT1CD19neg[,3:6]
```



## rs4841
```{r BAFrs4841}
snp150446963_c0 <- obj@loh$cluster0_baf[which(obj@loh$cluster0_baf$chr == "5" & obj@loh$cluster0_baf$position == "150446963"),]

snp150446963_c1 <- obj@loh$cluster1_baf[which(obj@loh$cluster1_baf$chr == "5" & obj@loh$cluster1_baf$position == "150446963"),]

snp150446963_c2_3_4_5 <- obj@loh$clusters_2_3_4_5_baf[which(obj@loh$clusters_2_3_4_5_baf$chr == "5" & obj@loh$clusters_2_3_4_5_baf$position == "150446963"),]

snp_freq_table <- matrix(data = c(snp150446963_c0$alt,snp150446963_c1$alt,snp150446963_c2_3_4_5$alt,snp150446963_c0$ref,snp150446963_c1$ref,snp150446963_c2_3_4_5$ref,snp150446963_c0$coverage,snp150446963_c1$coverage,snp150446963_c2_3_4_5$coverage,snp150446963_c0$baf,snp150446963_c1$baf,snp150446963_c2_3_4_5$baf), nrow = 3,ncol = 4 )

rownames(snp_freq_table) <- c("Cluster 0","Cluster 1","Clusters 2,3,4,5")
colnames(snp_freq_table) <- c("Alt","Ref","Coverage","BAF")
snp_freq_table

snp150446963_tumT1CD19neg <- loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6[which(loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$chr == "5" & loh_T1CD19n_in_tumor$cluster0_T1CD19neg_baf_mincov6$position == "150446963"),]
rownames(snp150446963_tumT1CD19neg) <- "Tumoral T1CD19negative cells"
snp150446963_tumT1CD19neg[,3:6]
```


```{r save the final CaSpER object, include=FALSE}
save(obj, file = paste0(WD_PATH, "output/obj_CaSpER.Robj"))
```

```{r session}
sessionInfo()
```
